### ISSUE TITLE
[Bug] Crash : Ankidroid crashes on checking media.

### ISSUE THREAD

**Issue by THElegend5** [ankidroid/Anki-Android#10892]
###### Reproduction Steps
- Updated from 2.15.6 (Play store ver.) to 2.16alpha58
 and when i opened a card i started getting "failed to load : failed to load ...."
![Screenshot_2022-04-16-17-34-50-744_com.ichi2.anki.png](https://user-images.githubusercontent.com/63903377/163674249-884f95df-96c1-44f1-8a0e-8a5fdc0edefd.png)
- So i decided to use check media which instantly crashed the app.
- After that i decided to force sync from anki web to ankidroid which also crashed the app.
- Decided to check if my library was corrupted for some reason or not by downloading a parallel of 2.15.6 on which it seemed to work perfectly, updated the parallel to alpha58 and it was also crashing.
- The alpha also doesn't seem to be detecting the backups neither is it creating them properly from what i can tell.
![Screenshot_2022-04-16-17-29-38-032_ru.zdevs.zarchiver.pro.png](https://user-images.githubusercontent.com/63903377/163674472-24c1d8ce-3c83-4239-b7cb-ed7f6b001482.png)
The only one which was good on 16/04 was made by the stable parallel.

- Just to confirm if the problem was from this alpha or from before this I decided to test on alpha 56 parallel and got the same issues uninstalled the parallel and installed alpha 55 parallel and there wasn't a crash this time, decided to update 55 to 56 and test again and there wasn't a crash, same with from 56 to 58.
- ~~Btw I did try to reinstall main alpha58 directly when i was originally getting the crash before the above point happened and the crashes still happened.~~ I thought i did but seems like i hadn't installed it directly and when I just did it seems to work fine on installing directly and this only happens if you update directly from 2.15.6 to 2.16 Alpha56/58/59.
- Have confirmed that it doesn't crash anymore on main after uninstalling 58 and installing 55 then 56 and 58 again after that.
- ~~Backup still don't seem to be being created properly but am not sure yet, so will confirm after I am.~~ It is detecting backups again and the backups are being created properly.

- Also half of this happened while writing this so it might seem like a story.

###### Expected Result

App checks media , directly syncs from anki web to ankidroid , Backup being created properly.

###### Actual Result

App crashes if you have updated directly to alpha 58 or installing alpha58 directly over current library.

###### Debug info
AnkiDroid Version = 2.16alpha58

Android Version = 11

Manufacturer = Xiaomi

Model = Redmi Note 9 Pro Max

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 11; Redmi Note 9 Pro Max Build/RKQ1.200826.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/100.0.4896.88 Mobile Safari/537.36

ACRA UUID = d9199133-a62c-4572-ad4f-899c4d0b5307

Scheduler = std2

Crash Reports Enabled = true

DatabaseV2 Enabled = true

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by david-allison** [comment]
Crash during check media:

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/d41c144d-1c3f-476d-8193-761f1e1375c4

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:415)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.NullPointerException: Attempt to get length of null array
at com.ichi2.libanki.Media._changes(Media.java:11)
at com.ichi2.libanki.Media._logChanges(Media.java:1)
at com.ichi2.libanki.Media.findChanges(Media.java:3)
at com.ichi2.async.CollectionTask$CheckMedia.task(CollectionTask.java:4)
at com.ichi2.async.CollectionTask$CheckMedia.task(CollectionTask.java:1)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:14)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$3.call(AsyncTask.java:394)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 4 more
```

**Comment by david-allison** [comment]
Hi @THElegend5, thanks for the report! 

Could you copy this into a new issue for the 'full sync' crash, so we can split out the two separate issues?

**Comment by THElegend5** [comment]
> Hi @THElegend5, thanks for the report! 
> 
> Could you copy this into a new issue for the 'full sync' crash, so we can split out the two separate issues?

Should i change anything or does copying everything works.

**Comment by david-allison** [comment]
copy everything please :)

**Comment by david-allison** [comment]
@Arthur-Milchior, this seems very likely related to https://github.com/ankidroid/Anki-Android/issues/10307

https://github.com/ankidroid/Anki-Android/blob/1a453679e62cabed8afe17ae4817d2ceeeae0d5e/AnkiDroid/src/main/java/com/ichi2/libanki/Media.java#L714

**Comment by THElegend5** [comment]
> copy everything please :)

Also the same title works or just limit that one to force sync.

**Comment by david-allison** [comment]
> Also the same title works or just limit that one to force sync.

Up to you: I /think/ these have the same root cause, but it's worth splitting them up just in case as it's two different crashes

**Comment by THElegend5** [comment]
> > Also the same title works or just limit that one to force sync.
> 
> Up to you: I /think/ these have the same root cause, but it's worth splitting them up just in case as it's two different crashes

Yeah was also thinking these might have the same cause the reason i reported them together , will just name them seperately to make it easier to tell which one is which.

**Comment by mikehardy** [comment]
I haven't logged it but had the same experience just yesterday: crash on full media check.
Additionally, if I logged out of ankiweb then logged in, I did not crash on the next sync but it did pop up an error during media sync saying trying to get length of null array or similar. Vague but main point is: I reproduce this as well

**Comment by dorrin-sot** [comment]
Hi can I take this? I think it's an issue of not having read access to external storage

**Comment by mikehardy** [comment]
But it's not. I definitely have "All files" permission allowed for the app. Feel free to look at it certainly! But it's not a permission issue

**Comment by THElegend5** [comment]
Yeah am sure too that it wasn't something to do with storage permission as all the cards were still there and opening as normal.
The cards which had media were just giving errors , media was still being displayed.

**Comment by dorrin-sot** [comment]
@mikehardy @THElegend5 But currently we don't have the "read access to external storage" permission. just "write access"

**Comment by mikehardy** [comment]
Read up on that permission in Android docs. Report findings

**Comment by THElegend5** [comment]
Just an update to the backups issue , it is detecting backups again and the backups are being created properly.

**Comment by dorrin-sot** [comment]
@mikehardy Looks like we wouldn't need to add it explicitly in manifest (it's implied with `WRITE_STORAGE_ACCESS`). and if the `READ_STORAGE_ACCESS` permission hasn't been grant, `listFiles()` can't find the folder and so returns null.

**Comment by mikehardy** [comment]
Correct, write contains read, so the declaration is fine. I definitely have granted the permission. So why is it a null media array and a crash 🤔

**Comment by dorrin-sot** [comment]
No clue 😕 (I'll look into why this would occur only after updating from v2.15.6 in the morning😅) but I tried logging `checkSelfPermission(this, READ_STORAGE_ACCEPTED) == PERMISSION_GRANTED` right before this line https://github.com/ankidroid/Anki-Android/blob/1a453679e62cabed8afe17ae4817d2ceeeae0d5e/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java#L1511 and it was false. I also saw that we are not asking the user for read access in https://github.com/ankidroid/Anki-Android/blob/1a453679e62cabed8afe17ae4817d2ceeeae0d5e/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java#L606
then, when I added it to the permission array in that line it got fixed.

**Comment by mikehardy** [comment]
Hmm. Perhaps it is Android itself being buggy about the permissions now that legacy storage is going away / gone for most everyone - we're pretty late on it. David just did #10894  - I'll do an alpha and perhaps that does it, or perhaps you're belt+suspenders approach of re-checking permissions / logging issues / re-requesting permissions will work around it

**Comment by mikehardy** [comment]
2.16alpha59 did not fix for me hmm
Needs further testing, i did not try full repro

![Screenshot_20220416-194420.jpg](https://user-images.githubusercontent.com/782704/163695657-2f8afa6d-0b6f-4051-bac8-053a510a99b8.jpg)

**Comment by david-allison** [comment]
> 2.16alpha59 did not fix for me hmm Needs further testing, i did not try full repro
> 
> ![Screenshot_20220416-194420.jpg](https://user-images.githubusercontent.com/782704/163695657-2f8afa6d-0b6f-4051-bac8-053a510a99b8.jpg)

Could you grab logs of folder/file ownership?

**Comment by THElegend5** [comment]
Some more stuff, I was wrong on one point above installing the Alpha directly doesn't seem to cause any problem which i found out when i was testing on Alpha59 , this seems to only be caused by updating from 2.15.6 to 2.16 Alpha 56/58/69.
Also i can reproduce everything if i update from 2.15.6 to Alpha 59. Also this probably isn't caused by the library itself because I was getting the same errors with my current library and the snapshot of my library before i decided to update while i was testing with Alpha59.

**Comment by dorrin-sot** [comment]
@mikehardy I logged 
```
Timber.i("storage permissions: read=%s write=%s",
     checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE),
     checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE));
```
in 58c5bd15e897bfb32a2badc80018729b0a7238d6 (v2.15.6) and master and the results were
- v2.15.6 : both were granted
- master : read was denied and write was granted
so maybe android changed how they ask storage permissions (if user accepted "write" permission android accepted "read" too)? or maybe when you're running `checkSelfPermission` on "read", android used to just check if "write" permission has been granted but recently they separated it?

**Comment by mikehardy** [comment]
I dunno - when I installed 2.16alpha60 it did not fix the problem, and it did not seem to *detect* the problem, but when I revoked storage permission, then re-added storage permission, it worked.

Really weird. As far as I can tell, the failure mode is that you can't read files *and* you can't tell you don't have storage permission somehow (that is: re-asking for it does not seem to matter). So the proper way to detect missing READ storage permission now might be to also check if you can read for media files? And if you have a null media files array, tell the user they need to revoke and re-grant storage permission? This all seems ridiculous

**Comment by dorrin-sot** [comment]
oh i added it to startup permissions so it needs a reinstall but didn't realize this was the issue before too🤦🏻 so sorry i'll add a permission check to "media check" too

**Comment by dorrin-sot** [comment]
@mikehardy is `DeclPicker::handleStartup` run at every startup or just at the first one?

**Comment by mikehardy** [comment]
But after an app update it definitely starts up clean and goes through the upgrade-the-app flow and everything ? I think it runs every time but I'm not 100%, I'd have to trace it out
